### PR TITLE
Refactor File Uploads

### PR DESCRIPTION
**Pull request by armstrongnate**

### CODE REVIEW

File: Core/Core.xcodeproj/xcshareddata/xcschemes/Core.xcscheme [hunk 1]
```diff
@@ -82,6 +87,11 @@
          </BuildableReference>
       </MacroExpansion>
       <AdditionalOptions>
+         <AdditionalOption
+            key = "NSZombieEnabled"
+            value = "YES"
+            isEnabled = "YES">
```
Comment (armstrongnate) [inline Core/Core.xcodeproj/xcshareddata/xcschemes/Core.xcscheme]: I need to turn off zombies.

File: Core/Core/API/API.swift [hunk 2]
```diff
@@ -97,6 +97,13 @@ public struct URLSessionAPI: API {
     @discardableResult
     public func uploadTask<R>(_ requestable: R, fromFile file: URL) throws -> URLSessionTask where R: APIRequestable {
         let request = try requestable.urlRequest(relativeTo: baseURL, accessToken: accessToken, actAsUserID: actAsUserID)
+        if let body = requestable.body {
+            let data = try requestable.encode(body)
+            let name = UUID().uuidString
+            let url = FileManager.default.temporaryDirectory.appendingPathComponent(name)
+            try data.write(to: url)
+            return urlSession.uploadTask(with: request, fromFile: url)
+        }
```
Comment (armstrongnate) [inline Core/Core/API/API.swift]: Need a test for this.

File: Core/Core/Database/CoreData/Models/File.swift [hunk 3]
```diff
@@ -36,14 +36,54 @@ final public class File: NSManagedObject {
...
+    public func prepareForSubmission(courseID: String, assignmentID: String) {
+        self.courseID = courseID
+        self.assignmentID = assignmentID
+    }
+
+    /// Marks the file as unsubmitted, removing reference to assignment.
+    public func markSubmitted() {
+        self.courseID = nil
+        self.assignmentID = nil
+    }
```
Comment (armstrongnate) [inline Core/Core/Database/CoreData/Models/File.swift]: Need tests for all these things.

File: Core/Core/Database/Persistence.swift [hunk 4]
```diff
@@ -46,9 +46,18 @@ extension PersistenceClient {
...
     }
+
+    public func first<T>(where key: String, equals value: CVarArg) -> T? {
+        return all(where: key, equals: value).first
+    }
+
+    public func all<T>(where key: String, equals value: CVarArg) -> [T] {
+        let predicate = NSPredicate(format: "%K == %@", key, value)
+        return fetch(predicate)
+    }
```
Comment (armstrongnate) [inline Core/Core/Database/Persistence.swift]: Need tests

File: Core/Core/Extensions/UIImageExtensions.swift [hunk 5]
```diff
@@ -28,12 +29,34 @@ extension UIImage {
...
+        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
+        let url = directory.appendingPathComponent(name, isDirectory: false).appendingPathExtension("png")
+        guard let data = pngData() else {
+            throw NSError.instructureError("Failed to save image")
+        }
+        if FileManager.default.fileExists(atPath: url.absoluteString) {
+            try FileManager.default.removeItem(at: url)
+        }
+        try data.write(to: url, options: .atomic)
+        return url
```
Comment (armstrongnate) [inline Core/Core/Extensions/UIImageExtensions.swift]: Need test

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  One or more files in Core are below the minimum test coverage 50%
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please run `yarn update-headers`. The following do not have the correct license header: 
* Core/CoreTests/Files/FileUploaderTests.swift
* TestsFoundation/TestsFoundation/Files/MockFileUploader.swift
* TestsFoundation/TestsFoundation/Fixtures/Model/FileUploadTargetFixture.swift
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  This pull request will not generate a release note.
  </td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-12078](https://instructure.atlassian.net/browse/MBL-12078)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Core** | 92.15% | 93.83% | -1.68%
Core/Core/AppEnvironment/AppGroup.swift | 0% | -- | --
Core/Core/Extensions/UIImageExtensions.swift | 15.79% | 100% | -84.21%
**Student** | 90.74% | 94.83% | -4.09%
**Teacher** | 100% | 100% | 0%
**React Native** | 92.54% | 92.54% | 0%
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Core/Core/Files/FileUploader.swift [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+        init(bundleID: String, appGroup: AppGroup?, userID: String, baseURL: URL, actAsUserID: String?) {
+            self.bundleID = bundleID
+            self.appGroup = appGroup
+            self.userID = userID
+            self.baseURL = baseURL
+            self.actAsUserID = actAsUserID
+        }
+
+        init?(string: String) {
```
Comment (vanwagonet) [inline Core/Core/Files/FileUploader.swift]: `init?(identifier: String)` to make a clear connection with the `identifier` below.

File: Core/Core/Files/FileUploader.swift [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        self.database = database
+        let configuration = URLSessionConfiguration.background(withIdentifier: identifier)
+        configuration.sharedContainerIdentifier = appGroup?.rawValue
+        super.init()
+        backgroundSession = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
+    }
+
+    public func upload(_ file: File, context: FileUploadContext, callback: @escaping (Error?) -> Void) {
+        Logger.shared.log()
+        let defaultError = NSError.instructureError(NSLocalizedString("Could not upload file", comment: ""))
```
Comment (vanwagonet) [inline Core/Core/Files/FileUploader.swift]: I've seen some `NSError`s and `enum: Error`s. Do we want to try to make that consistent, or do they have different uses?

File: Core/Core/Files/FileUploader.swift [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+        }
+    }
+}
+
+// MARK: - Submissions
+extension FileUploader {
+    private func submit(assignmentID: String, courseID: String) {
+        performAndWait { context in
+            let files: [File] = context.all(where: #keyPath(File.assignmentID), equals: assignmentID)
+            let ready = files.allSatisfy { $0.isUploaded } == true
```
Comment (vanwagonet) [inline Core/Core/Files/FileUploader.swift]: is the `== true` redundant here? `allSatisfy` returns a `Bool`.

File: Core/Core/AppEnvironment/AppGroup.swift [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+public enum AppGroup: String, Codable {
+    case student = "group.com.instructure.icanvas"
+
+    public var url: URL {
+        guard let url = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: rawValue) else {
+            preconditionFailure("app group not found \(rawValue)")
+        }
+        return url
+    }
```
[the remark this answers is not in the document]
Comment (armstrongnate) [reply Core/Core/AppEnvironment/AppGroup.swift]: Good call. I removed AppGroup and use that instead.

File: Core/Core/Database/Persistence.swift (hunk 4 above)
Comment (armstrongnate) [reply Core/Core/Database/Persistence.swift]: Done

File: Core/Core/Files/FileUploader.swift (hunk 6 above)
Comment (armstrongnate) [reply Core/Core/Files/FileUploader.swift]: Done

File: Core/Core/Files/FileUploader.swift (hunk 8 above)
Comment (armstrongnate) [reply Core/Core/Files/FileUploader.swift]: Ah yes. This used to be optional. Fixed.

**Comment by armstrongnate** [comment]
The UI tests failed but I don't think it's related. `testSubmitOnlineUpload()` did pass..

**Pull request merged**